File: python/src/can_construct.py
# Check if it's possible to create target string by using words from wordbank

from common import time_execution


@time_execution
def can_construct_rec(target: str, word_bank: list[str]) -> bool:
    def helper(target: str):
        if target == '':
            return True
        for word in word_bank:
            if target.startswith(word):
                if helper(target.lstrip(word)):
                    return True
        return False

    return helper(target)


@time_execution
def can_construct_rec_memo(target: str, word_bank: list[str]) -> bool:
    memo = {'': True}

    def helper(target: str):
        if target in memo:
            return memo[target]

        for word in word_bank:
            if target.startswith(word):
                if (res := helper(target.lstrip(word))):
                    memo[target] = res
                    return res

        memo[target] = False
        return False

    return helper(target)


@time_execution
def can_construct_rec_v2(target: str, word_bank: list[str]) -> bool:
    word_bank_len = len(word_bank)

    def helper(target: str, idx=0):
        if idx == word_bank_len:
            return False
        if target == '':
            return True

        if target.startswith(word_bank[idx]):
            return helper(target.lstrip(word_bank[idx]), 0) \
                or helper(target, idx+1)
        else:
            return helper(target, idx+1)

    return helper(target)


@time_execution
def can_construct_rec_v2_memo(target, word_bank):
    word_bank_len = len(word_bank)
    memo = {'': True}

    def helper(target, idx=0):
        if idx == word_bank_len:
            return False
        if target in memo:
            return memo[target]

        if target.startswith(word_bank[idx]):
            memo[target] = helper(target.lstrip(
                word_bank[idx]), 0) or helper(target, idx + 1)
            return memo[target]
        else:
            memo[target] = helper(target, idx + 1)
            return memo[target]

    return helper(target)


print(can_construct_rec("enterapotentpot", [
      "a", "p", "ent", "enter", "ot", "o", "t"]))
print(can_construct_rec_memo("enterapotentpot", [
      "a", "p", "ent", "enter", "ot", "o", "t"]))

print(can_construct_rec_v2("enterapotentpot", [
      "a", "p", "ent", "enter", "ot", "o", "t"]))
print(can_construct_rec_v2_memo("enterapotentpot", [
      "a", "p", "ent", "enter", "ot", "o", "t"]))

File: php/src/can_construct.php
<?php

function can_construct_dp(string $target, array $word_bank): bool
{
    $dp = array_fill(0, strlen($target) + 1, null);
    $dp[0] = "";
    for ($i = 0; $i <= strlen($target); $i++) {
        if ($dp[$i] !== null) {
            foreach ($word_bank as $word) {
                $new_word = $dp[$i] . $word;
                if ($new_word == $target)
                    return true;
                if (!str_starts_with($target, $new_word))
                    continue;

                $dp[strlen($new_word)] = $new_word;
            }
        }
    }
    return false;
}

function can_construct_dp_v2(string $target, array $word_bank): bool
{
    $dp = array_fill(0, strlen($target) + 1, false);
    $dp[0] = true;
    for ($i = 0; $i <= strlen($target); $i++) {
        if ($dp[$i]) {
            foreach ($word_bank as $word) {
                if ($word == substr($target, $i, strlen($word)))
                    $dp[$i + strlen($word)] = true;
            }
        }
    }
    return end($dp);
}

function can_construct_rec(string $target, array $word_bank): bool
{
    $helper = function (int $idx = 0) use ($target, $word_bank, &$helper) {
        if ($idx == strlen($target))
            return true;
        foreach ($word_bank as $word) {
            if (str_starts_with(substr($target, $idx), $word)) {
                if ($helper($idx + strlen($word))) {
                    return true;
                }
            }
        }
        return false;
    };

    return $helper();
}


function can_construct_rec_v2(string $target, array $word_bank): bool
{
    $helper = function (string $target, int $idx = 0) use ($word_bank, &$helper) {
        if ($idx == count($word_bank))
            return false;
        if ($target == "")
            return true;

        if (str_starts_with($target, $word_bank[$idx]))
            return $helper(ltrim($target, $word_bank[$idx]), 0) || $helper($target, $idx + 1);

        return $helper($target, $idx + 1);
    };

    return $helper($target);
}

function can_construct_rec_v3(string $target, array $word_bank): bool
{
    $helper = function (string $target) use ($word_bank, &$helper) {
        if (empty($target)) {
            return true;
        }
        foreach ($word_bank as $word) {
            if (str_starts_with($target, $word)) {
                if ($helper(ltrim($target, $word))) {
                    return true;
                }
            }
        }
        return false;
    };

    return $helper($target);
}

function can_construct_rec_memo(string $target, array $word_bank): bool
{
    $memo = ["" => true];
    $helper = function (string $target) use ($word_bank, $memo, &$helper) {
        if (isset($memo[$target])) {
            return true;
        }
        foreach ($word_bank as $word) {
            if (str_starts_with($target, $word)) {
                $memo[$target] = $helper(ltrim($target, $word));
                if ($memo[$target]) {
                    return true;
                }
            }
        }
        $memo[$target] = false;
        return false;
    };

    return $helper($target);
}